### PR TITLE
BUG: Validate inferred dtype in sparse constructors

### LINKED ISSUE #25272: BUG: sparse: check for valid dtype in all constructor use patterns
While reviewing the recent deprecation PR #25245, I found a number of cases where we don't check for a valid dtype when constructing sparse arrays. For example,
```python
sp.sparse.csr_array((np.array([None]), np.array([0]), np.array([0, 1]))) # allows dtype='object' 
sp.sparse.csr_array((np.array([1], dtype=np.float16), np.array([0]), np.array([0, 1]))) # allows dtype='float16'
```
 
These constructions are allowed and seem to work, but then we raise a TypeError when any operation on the sparse array checks the dtype. For example, `str(A)` raises a TypeError indicating an unsupported dtype.

I propose that we make a follow-up to #20754 which added checks to each `__init__` call, but seems to have missed cases when `dtype` is not specified.  That is, when `dtype=` is specified we check whether that is supported. But when `dtype=None` and a dtype is inferred from input arrays we don't check that the resulting dtype is supported.  

I think the `_sputils.getdtype` helper function should be sufficient. But we will need to find all the ways it might arise -- probably with a test over all formats using `dtype=None` and input data of 'float16' and/or 'object'.

### PR DESCRIPTION
**Pull request by rahulrathnavel**
Constructors for sparse formats now properly validate the dtype inferred from the input data (e.g. `object`, `float16`) to prevent late-stage errors.

#### Reference issue
Closes gh-25272

#### What does this implement/fix?
Previously, when `dtype=None` was passed to sparse array/matrix constructors, the inferred data type was not validated, allowing unsupported types to be silently initialized. This PR ensures inferred types are strictly validated across all formats.

Specific changes made:
* Updated `_sputils.getdtype` to raise a `TypeError` instead of a `ValueError` for unsupported dtypes to match expected behavior.
* Replaced raw `np.array` initialization with the `getdata` utility in `_compressed.py` (CSR/CSC) and `_dia.py` (DIA) to ensure inferred types correctly pass through validation.
* Updated existing tests in `test_base.py` and `test_sputils.py` to expect the new `TypeError`.
* Added `test_unsupported_dtypes_in_constructors` to `test_construct.py` to verify this behavior across all sparse formats.

#### Additional information
By leveraging the existing `getdata` utility across the remaining formats, this fix plugs the validation gap without adding unnecessary or complex `if/else` checks, keeping the codebase clean and standard.

#### AI Generation Disclosure
I used an AI coding assistant (Gemini) to help draft the syntax and format the test functions. All architectural logic, testing, and implementation decisions were manually driven, reviewed, and verified by me to ensure they meet SciPy's standards.

### CODE REVIEW

**Comment by rahulrathnavel** [comment]
Hi @perimosocordiae @dschult i have updated getdtype to check newdtype.type instead of the raw dtype, which allows it to safely ignore byteorder/endianness (like >f8) while still strictly blocking unsupported core types like object or float16,ensures Validate inferred dtype in sparse constructors.
 Let me know if this looks good or if any further tweaks are needed before merging!😄

**Review by dschult — COMMENTED** [review]
Thanks for this -- I've got a number of comments -- most of them are reasonable choices but not what fits in scipy.

- we want ValueError, not TypeError.  A TypeError refers to an invalid Python type, which is not the same as a numpy dtype (dtype is the data type--closer to a C type than a python type).  Anyway, the type of the input will be a numpy array. The dtype will be something like np.int32. So we want all those to raise ValueErrors.
- In _sputils.py  getdtype we want to check newdtype instead of newdtype.type. We don't want to ignore endian... Or at least the behavior should be the same as it currently is.  Treatment of endian-ness should stay the same, if I understand correctly.
- This one is just personal preference for me: I'd rather keep the `np.array` instead of switching to `getdata`. Then apply `getdtype` to the dtype.  Something like: `np.array(arg1[0], dtype=getdtype(dtype), copy=copy)`.  My reasoning is that `getdata` hides the building of a numpy array, and only adds a call to getdtype which we can just do in the call itself without this additional utility function call.
- The tests should not be in test_construct.py... though that name could seem right. The `construct.py` module relates to construction functions like `random_array` and `vstack`, whereas the "constructor" functions are the class constructors. So I think this new test belongs in `test_base.py` in the `TestCommon` class... how about right after `test_invalid_shapes` (~line 705).  This test class is included in all of the formats, so you no longer need the loop over `sp_format`. Instead `sp_format` use   `self.spcreator` to create the object. And the if clause with sp_format has to be something like: `if self.sp_creator._format in ('csc', 'csr'):`.  I guess we could put the test into each of the `TestCSR` classes and avoid the if-clauses. But then they are spread out across the tests and are thus harder to ensure that all possibilities are      checked. So I would keep them together.
- minor preference for the term `some_unsupported_dtypes` rather than `unsupported _dtypes` so it is clear that we aren't testing all possibilities.

Let me know if/when any of this isn't clear.  Thanks!

**Comment by rahulrathnavel** [comment]
@dschult @perimosocordiae  all  CI checks are passed! The adjustment to let np.array() handle the list inputs first worked perfectly and cleared up all the pipeline crashes. Let me know if this looks good to merge now else  just suggest me what should i need to do next!😄

File: scipy/sparse/tests/test_base.py [hunk 1]
```diff
@@ -703,6 +703,13 @@ def test_invalid_shapes(self):
         assert_raises(ValueError, self.spcreator, (3,-1))
         assert_raises(ValueError, self.spcreator, (-1,-1))
 
+    def test_unsupported_dtypes_in_constructors(self):
+        some_unsupported_dtypes = [np.float16, object]
+        for dt in some_unsupported_dtypes:
+            data = np.array([[1]], dtype=dt)
+            with assert_raises(ValueError, match="does not support dtype"):
+                self.spcreator(data)
```
Comment (j-bowhay) [inline scipy/sparse/tests/test_base.py:711]:
```suggestion
    @pytest.mark.parametrize("dtype", [np.float16, object])
    def test_unsupported_dtypes_in_constructors(self, dtype):
        data = np.array([[1]], dtype=dtype)
        with assert_raises(ValueError, match="does not support dtype"):
            self.spcreator(data)
```

**Comment by rahulrathnavel** [comment]
Hi@j-bowhay i was just applied your suggestion and thanks for your suggestion on using parameterize.

It looks like the CI checks immediately failed across the board right after the commit.so should i need to fix since i think pytest might not be imported at the top of test_base.py,causing a NameError when the decorator tries to runthat i guess so let me know what you think and what i should do next!
Thanks.

**Comment by rahulrathnavel** [comment]
@j-bowhay i tried doing that manually and ran into the same errors so i just rollbacked to your suggested commits😅. Now i understand that i need your guidance/suggestions to move forward from here. Otherwise, if its okay to merge without addressing these failing cases (if not affects main) then i am happy either way!

File: scipy/sparse/_compressed.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, *, maxprint=None):
                     self.indices = np.array(indices, copy=copy, dtype=idx_dtype)
                     self.indptr = np.array(indptr, copy=copy, dtype=idx_dtype)
                     self.data = np.array(data, copy=copy, dtype=dtype)
+                    getdtype(self.data.dtype.type)
```
Comment (dschult) [inline scipy/sparse/_compressed.py]: I think combining getdtype with the array creation is cleaner. Do you?
```suggestion
self.data = np.array(data, copy=copy, dtype=getdtype(dtype, data))
```

File: scipy/sparse/_dia.py [hunk 3]
```diff
@@ -59,6 +59,7 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, *, maxprint=None):
                     if not copy:
                         copy = copy_if_needed
                     self.data = np.atleast_2d(np.array(arg1[0], dtype=dtype, copy=copy))
+                    getdtype(self.data.dtype.type)
```
Comment (dschult) [inline scipy/sparse/_dia.py]:
```suggestion
self.data = np.atleast_2d(np.array(arg1[0], dtype=getdtype(dtype, arg1[0]), copy=copy))
```

**Review by dschult — COMMENTED** [review]
Yeah--  the `test_base.py` machinery for e.g. non-canonical data doesn't work with pytest.mark.parametrize. When I looked into that a couple years ago it looked like it would require a major rewrite of test_base.py. So let's just leave the loop in the method and don't parametrize.

I think both remaining changes can be done when setting up the array instead of on a separate line. Then the dtype check is done first -- and both steps are done on the same line -- maybe making it easier to read.

**Comment by rahulrathnavel** [comment]
Hi @dschult , i applied your suggestions exactly as requested.
I just had to manually break the line in _dia.py into a couple of lines because the i think  GitHub UI button made it 107 characters long, which tripped the lint error.
Let me know if this looks clean to you now or if you have any other suggestions in mind!
(@j-bowhay if need to do something with those failing checks happy to do the suggestions further)thanks in advance.

File: scipy/sparse/_dia.py [hunk 4]
```diff
@@ -58,7 +58,9 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False, *, maxprint=None):
                         raise ValueError('expected a shape argument')
                     if not copy:
                         copy = copy_if_needed
-                    self.data = np.atleast_2d(np.array(arg1[0], dtype=dtype, copy=copy))
+                    self.data = np.atleast_2d(
+                        np.array(arg1[0], dtype=getdtype(dtype, arg1[0]), copy=copy)
+                    )
```
Comment (dschult) [inline scipy/sparse/_dia.py]: Rats -- sorry that my suggestion was ignoring the fact that the input might not be an array yet.
The original ordering of creating the array and then checking the dtype is probably best without a rewrite of getdtype to handle `a` not being an array.  Maybe we can add a comment like
`getdtype(self.data.dtype)  # check that dtype is supported`

Same is true for the `_compressed.py` case. But is there a reason to use `self.data.dtype.type` instead of `self.data.dtype` there?  We don't want to ignore the endian stuff if I understand correctly.

**Comment by rahulrathnavel** [comment]
Thanks @dschult good to hear it from you really as i got very confused and felt bit lost, now the things are better and clear to me now😅i hope it will work too!

File: scipy/sparse/_dia.py (hunk 4 above)
Comment (dschult) [reply scipy/sparse/_dia.py]: Well -- it looks like there **is** a reason to use `self.data.dtype.type` instead of `self.data.dtype`...
I'm really sorry that I keep steering you in the wrong direction. 

I've looked into the history some, I've looked into what sparsetools operations this might affect. I've tried to figure out whether allowing big endian on a little endian-native machine works. They all seem to work. So here's my current thinking:  We should change `getdtype()` to check `if newdtype.type not in supported_dtypes:`. Then the potentially confusing part is in one place instead of two. The calls will input `self.data.dtype`, which is pretty clear. And we should add a comment on a line before the `if` line with something like:  `# check newdtype.type to avoid raising for endians which get fixed later`

Let's hope we get all green checks!   Thanks!

**Comment by rahulrathnavel** [comment]
@dschult @j-bowhay @perimosocordiae Hi all,
It’s been a while ,i would really like to fix this issue😅 any suggestions or guidance on how to implement the fix would be much appreciated.
If anyone have any ideas or feedback just anything kindly ping me let me know so i can try it out . I will be very happy to try them out.
Thanks in advance!

**Comment by dschult** [comment]
The issue is whether to ignore endian-ness when checking the dtype. I've looked into the history some, I've looked into what sparsetools operations this might affect. I've tried to figure out whether allowing big endian on a little endian-native machine works. So here's my current thinking (this backtracks on some suggestions I made earlier -- sorry for the misdirection: 

We should change getdtype() to check `if newdtype.type not in supported_dtypes:` Then the potentially confusing use of dtype.type is in one place instead of two. The calls input `self.data.dtype`, which is pretty clear. And we can add a comment on the line before the "if" line with something like: `# check newdtype.type to avoid raising for endians which get fixed later`

Let's hope we get all green checks! Thanks @rahulrathnavel

**Comment by rahulrathnavel** [comment]
> The issue is whether to ignore endian-ness when checking the dtype. I've looked into the history some, I've looked into what sparsetools operations this might affect. I've tried to figure out whether allowing big endian on a little endian-native machine works. So here's my current thinking (this backtracks on some suggestions I made earlier -- sorry for the misdirection:
> 
> We should change getdtype() to check `if newdtype.type not in supported_dtypes:` Then the potentially confusing use of dtype.type is in one place instead of two. The calls input `self.data.dtype`, which is pretty clear. And we can add a comment on the line before the "if" line with something like: `# check newdtype.type to avoid raising for endians which get fixed later`
> 
> Let's hope we get all green checks! Thanks @rahulrathnavel

@dschult no need to worry about anything i am really happy that your travelling and supporting all through this journey!😄

Just now pushed your suggestions! so now it looks good to me so i hope lets see the reviewers will say the same thing and merge it soon!😄 @j-bowhay @perimosocordiae

File: scipy/sparse/_sputils.py
Comment (perimosocordiae) [inline scipy/sparse/_sputils.py:116]: The docs here indicate that a type error could be raised rather than a value error. I tend to agree with Dan that a value error makes sense here, but for the sake of backwards compatibility, it might be safer to continue raising type errors.

Comment (rahulrathnavel) [reply scipy/sparse/_sputils.py:116]: @perimosocordiae i believe I have implemented the changes in as you suggested way😅so kindly verify it and if further changes needed let me know it if nothing else needed the. All checks have been passed already so I think this would be perfect to get merged! Thank you.

File: scipy/sparse/_sputils.py [hunk 5]
```diff
@@ -127,11 +127,12 @@ def getdtype(dtype, a=None, default=None):
             if default is not None:
                 newdtype = np.dtype(default)
             else:
-                raise TypeError("could not interpret data type") from e
+                raise ValueError("could not interpret data type") from e
```
Comment (dschult) [inline scipy/sparse/_sputils.py]: Good point @perimosocordiae about not introducing backward incompatible changes with the error type.

Let's switch this back to `TypeError`...

Comment (rahulrathnavel) [reply scipy/sparse/_sputils.py]: @dschult as I have implemented the changes suggested by @perimosocordiae already and all CI checks are now green so just waiting for him to review and merge the code 😄thank you both for the support!

Comment (perimosocordiae) [reply scipy/sparse/_sputils.py]: @dschult  is correct here, this should remain a TypeError.

**Review by perimosocordiae — APPROVED** [review]
Approved pending one more error class change.

**Review by dschult — APPROVED** [review]
This looks good!  Thanks @rahulrathnavel !!

**Comment by rahulrathnavel** [comment]
> This looks good! Thanks @rahulrathnavel !!

😁 Happy' to hear this from you and really special thanks to you for guiding me all through this journey! Hoping to work on more fixes soon😄  Felt like a wonderful 🎢 roller coster ride 😅! 
@dschult @j-bowhay @perimosocordiae 
thank you all!
